src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs

```diff
@@ -30,63 +30,131 @@
+            // load into the compiler or directory load context.
+            //
+            // Not only is this bad behavior it also pollutes future test results.
+            var count = AssemblyLoadContext.Default.Assemblies.Count();
+            using var fixture = new AssemblyLoadTestFixture();
+            using var tempRoot = new TempRoot();
+            var loader = shadowLoad
+                ? new ShadowCopyAnalyzerAssemblyLoader(alc, tempRoot.CreateDirectory().Path)
+                : new DefaultAnalyzerAssemblyLoader(alc);
+            try
+            {
+                DefaultAnalyzerAssemblyLoaderTests.InvokeTestCode(loader, fixture, typeName, methodName);
+            }
+            finally
+            {
+                testOutputHelper($"Test fixture root: {fixture.TempDirectory.Path}");
+
+                foreach (var context in loader.GetDirectoryLoadContextsSnapshot())
+                {
+                    testOutputHelper($"Directory context: {context.Directory}");
+                    foreach (var assembly in context.Assemblies)
+                    {
+                        testOutputHelper($"\t{assembly.FullName}");
+                    }
+                }
+
+                if (loader is ShadowCopyAnalyzerAssemblyLoader shadowLoader)
+                {
+                    testOutputHelper($"Shadow loader: {shadowLoader.BaseDirectory}");
+                }
+
+                testOutputHelper($"Loader path maps");
+                foreach (var pair in loader.GetPathMapSnapshot())
+                {
+                    testOutputHelper($"\t{pair.OriginalAssemblyPath} -> {pair.RealAssemblyPath}");
+                }
+
+                Assert.Equal(count, AssemblyLoadContext.Default.Assemblies.Count());
+            }
         }
+    }
+
 #else
-        private sealed class InvokeUtil : MarshalByRefObject
+
+    public sealed class InvokeUtil : MarshalByRefObject
+    {
+        public void Exec(ITestOutputHelper testOutputHelper, bool shadowLoad, string typeName, string methodName)
         {
-            public void Exec(string typeName, string methodName)
+            using var fixture = new AssemblyLoadTestFixture();
+            using var tempRoot = new TempRoot();
+            var loader = shadowLoad
+                ? new ShadowCopyAnalyzerAssemblyLoader(tempRoot.CreateDirectory().Path)
+                : new DefaultAnalyzerAssemblyLoader();
+
+            try
             {
-                try
+                DefaultAnalyzerAssemblyLoaderTests.InvokeTestCode(loader, fixture, typeName, methodName);
+            }
+            catch (TargetInvocationException ex) when (ex.InnerException is XunitException)
+            {
+                var inner = ex.InnerException;
+                throw new Exception(inner.Message + inner.StackTrace);
+            }
+            finally
+            {
+                testOutputHelper.WriteLine($"Test fixture root: {fixture.TempDirectory.Path}");
+
```

Comment:
One thing I am wondering is, is this just a test baseline which should be recorded and turned into a failure when things change? Or is it much too brittle for that?

Reply:
It's too brittle for that. The shadow copy base directory changes on every run of the test. Also the set of test assemblies, particularly in the default contexts, change with some regularity. Consider that every change of a version, NuPkg update, etc ... change the output. 

For the most part I've found just the extra data here is enough to diagnose what is going wrong. Even in CI situations.